### ISSUE TITLE
PATCH /v2/entities/ID/attrs error doesn't provide info about offending attributes

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#2933]
Error when `POST /v2/entities/ID/attrs?options=append` fails:

```
{"error":"Unprocessable","description":"one or more of the attributes in the request already exist: [ temperature, pressure ]"}
```

Error when `PATCH /v2/entities/ID/attrs` fails:

```
{"error":"NotFound","description":"The entity does not have such an attribute"}
```

The second one should be probably be more explicit, as the first one. Not sure, I don't remember if this is the way it is due to some reasons (maybe some NGSIv2 spec aspect?)

**Comment by rajeswar-NEC** [comment]
Hi @fgalan 

I  would like to work on this issue, please let me know if no one is working on this.
As per my understanding, the error message string of API “PATCH /v2/entities/ID/attrs” in case of failure, need to be updated. 
Please confirm.

**Comment by rajeswar-NEC** [comment]
Hi @fgalan 

According to NGSIv2 Spec, in case of failure the API PATCH /v2/entities/ID/attr  returned response error string {“error” : “NotFound”}  (if the resource identified by the request is not found then “NotFound” is returned)

So, I have updated returned response description string with more details as below:
{"error":"NotFound","description":"action: UPDATE - entity: [Room1, Thing]The entity does not have such an attribute - offending attribute: temperatur"}

Please confirm as per issue is it I have to change returned response of description string and also please confirm my changes and what type of string details  it should return.

please find the below screenshots for more details

I have registered entity name “Room1” with  "pressure" and “temperature”  attributes 

![Screenshot (37)](https://user-images.githubusercontent.com/55490519/91034324-7a19a900-e622-11ea-802d-6f40f713912e.png)


Case 1: I have updated entity "Room1" attribute values with PATCH API, I have tried with non-existing attribute name (“pressur”) and returned description with offending attribute name.

![Screenshot (35)](https://user-images.githubusercontent.com/55490519/91034302-7128d780-e622-11ea-96c6-41714152bfd4.png)


Case 2: I have updated entity "Room1" attribute values with PATCH API, I have tried with non-existing attribute name (“temperatur”) and returned description with offending attribute name.

![Screenshot (33)](https://user-images.githubusercontent.com/55490519/91033840-a8e34f80-e621-11ea-9f9e-f456982b8264.png)

**Comment by fgalan** [comment]
> Please confirm as per issue is it I have to change returned response of description string and also please confirm my changes and what type of string details it should return.

Looking to the [NGSIv2 spec](http://telefonicaid.github.io/fiware-orion/api/v2/stable/) section "Error responses" I would't say that errors to `POST /v2/entities/ID/attrs` due to non-existing attribute fall in this case:

> Errors due to the request plus state combination but not exclusively from the request results in `Unprocessable (422)`.

Thus the errors for case 1 and 2 should be the following ones (ideally):

```
{"error":"Unprocessable","description":"one or more of the attributes in the request do not exist: [ presur ]"}
{"error":"Unprocessable","description":"one or more of the attributes in the request do not exist: [ temperatur ]"}
```

What do you think?

**Comment by rajeswar-NEC** [comment]
Hi @fgalan 
As per my understanding error responds  for case 1 & case 2 should be following

{"error":"NotFound","description":"one or more of the attributes in the request do not exist: [ presur ]"}
{"error":"NotFound","description":"one or more of the attributes in the request do not exist: [ temperatur ]"}

please let me know your opinion.

**Comment by fgalan** [comment]
> Hi @fgalan
> As per my understanding error responds for case 1 & case 2 should be following
> 
> {"error":"NotFound","description":"one or more of the attributes in the request do not exist: [ presur ]"}
> {"error":"NotFound","description":"one or more of the attributes in the request do not exist: [ temperatur ]"}
> 
> please let me know your opinion.

I have provided a rationale, based on NGSIv2 specification, that tells otherwise (see [my comment above](https://github.com/telefonicaid/fiware-orion/issues/2933#issuecomment-687084186)). Could you elaborate on which is bases your understanding, please?

**Comment by Anjali-NEC** [comment]
Hi @fgalan sir,
Due to medical reason @rajeswar-NEC is not able to work on this issue, so I am working on this issue. 
I have raised new PR for the same.
Please have a look PR https://github.com/telefonicaid/fiware-orion/pull/3900 

https://github.com/telefonicaid/fiware-orion/pull/3686#discussion_r487855221
As per your comment we need to show more than one non existing attribute in the form of array.
For this we need to get all the non existing attributes and need to arrange into an array in `src/lib/mongoBackend/MongoCommonUpdate.cpp` in function `updateContextAttributeItem`
Currently only the last non existing attribute is showing in error even there are more non existing attributes are exist in the request. I am trying to get the other non existing attributes other than targetAtrr.
I am trying to find other attributes using  (targetAttr-> previousValue)->getName(). but I am not able to get the next
targetAtrr name.

Can you please suggest where other non existing attribute are stored.

**Comment by fgalan** [comment]
Maybe you should use an approach similar to the one used for detecting existing attribute in strict append cases (e.g. `POST /v2/entities/E/attrs?options=append`). The call to `updateEntity()` (in `processContextElement()` at MongoCommonUpdate.cpp) is this way:

```
    notifSent = updateEntity(results[ix],
                             action,
                             tenant,
                             servicePathV,
                             xauthToken,
                             eP,
                             responseP,
                             &attributeAlreadyExistsError,
                             &attributeAlreadyExistsList,
                             forcedUpdate,
                             apiVersion,
                             fiwareCorrelator,
                             notifStartCounter,
                             ngsiV2AttrsFormat);
```

Note the `attributeAlreadyExistsError` and `attributeAlreadyExistsList` parameters. They are later used (in `processContextElement()`) to generate the error response: 

```
  if (attributeAlreadyExistsError == true)
  {
    std::string details = "one or more of the attributes in the request already exist: " + attributeAlreadyExistsList;
    buildGeneralErrorResponse(eP, NULL, responseP, SccBadRequest, details);
    responseP->oe.fill(SccInvalidModification, details, "Unprocessable");
  }
```

How `attributeAlreadyExistsList` is filled? Have a look to `updateEntity()` function and you would find this pre-condition checking for strict append case:

```
  if (action == ActionTypeAppendStrict)
  {
    for (unsigned int ix = 0; ix < eP->attributeVector.size(); ++ix)
    {
      if (attrs.hasField(eP->attributeVector[ix]->name))
      {
        alarmMgr.badInput(clientIp, "attribute already exists");
        *attributeAlreadyExistsError = true;

        //
        // This attribute should now be removed from the 'query' ...
        // processContextAttributeVector looks at the 'skip' field
        //
        eP->attributeVector[ix]->skip = true;

        // Add to the list of existing attributes - for the error response
        if (*attributeAlreadyExistsList != "[ ")
        {
          *attributeAlreadyExistsList += ", ";
        }
        *attributeAlreadyExistsList += eP->attributeVector[ix]->name;
      }
    }
    *attributeAlreadyExistsList += " ]";
  }
...
    if (!(attributeAlreadyExistsError && (action == ActionTypeAppendStrict)))
    {
      // Note that CER generation in the case of attributeAlreadyExistsError has its own logic at
      // processContextElement() function so we need to skip this addition or we will get duplicated
      // CER
      responseP->contextElementResponseVector.push_back(cerP);
    }
...
```

So maybe you could use a similar approach, in this case for `ActionTypeUpdate` action type. You could add a new pair of parameters for that in the `updateEntity()` function (let's say `attributeNotExistingList` and `attributeNotExistingError`) or, even better, try to reuse the existing ones for the two cases (`ActionTypeAppendStrict` and `ActionTypeUpdate`).

I hope this information can be helpful.

**Comment by Anjali-NEC** [comment]
Hi @fgalan Sir,

As per your suggestion I have added two new parameters attributeNotExistingList and attributeNotExistingError smiliar to attributeAlreadyExistsError and attributeAlreadyExistsList parameters and used these new parameters for getting not existing attribute list and error for the same. 

I have also add modified updateEntity() function at MongoCommonUpdate.cpp and added logic for ActionTypeUpdate action type in my PR #https://github.com/telefonicaid/fiware-orion/pull/3900 

Due to this several existing test cases are failing. Please guide me further.

**Comment by fgalan** [comment]
Fixed by PR https://github.com/telefonicaid/fiware-orion/pull/3900

**Comment by naveennec** [comment]
@fgalan @kzangeli  I hope this issue is closed but as per yesterday TSC meeting Status I got confused

"ONGOING - PATCH request fixes (fix for issue #2933) #3686 @ rajeswar-NEC  [NEC] (X) - stalled? (+1 Year)"

I hope PR 3686 is no more relevant right when another PR(3900) is merged and issue is closed.
 Can you please cancel/close the pull request(3686) because as per comment number 
[Rajeswara is no more working with NEC and I don't have rights to close the PR.](https://github.com/telefonicaid/fiware-orion/issues/2933#issuecomment-889116147)  ,Anjali has already shared the concern.
Thanks

**Comment by fgalan** [comment]
> @fgalan @kzangeli I hope this issue is closed but as per yesterday TSC meeting Status I got confused
> 
> "ONGOING - PATCH request fixes (fix for issue #2933) #3686 @ rajeswar-NEC [NEC] (X) - stalled? (+1 Year)"
> 
> I hope PR 3686 is no more relevant right when another PR(3900) is merged and issue is closed. Can you please cancel/close the pull request(3686) because as per comment number [Rajeswara is no more working with NEC and I don't have rights to close the PR.](https://github.com/telefonicaid/fiware-orion/issues/2933#issuecomment-889116147) ,Anjali has already shared the concern. Thanks

You are right. PR #3900 (which seems to be an evolved version of PR #3686) solves issue #2933. It seems that we forgot to close PR #3686 when we merged PR #3900. I have closed it now.

Thanks for the clarifications and sorry for the confusion!
